go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -3015,7 +3176,201 @@ func (s *VtctldServer) ValidateShard(ctx context.Context, req *vtctldatapb.Valid
+	if err != nil {
+		resp.Results = append(resp.Results, fmt.Sprintf("unable to get reference version of first shard's primary tablet: %v", err))
+		return &resp, nil
+	}
+
+	var validateVersionKeyspaceResponseMutex sync.Mutex
+
+	for _, shard := range shards {
+		validateVersionKeyspaceResponseMutex.Lock()
+		shardResp := vtctldatapb.ValidateShardResponse{
+			Results: []string{},
+		}
+
+		var (
+			validateShardResponseMutex sync.Mutex
+			tabletWaitGroup            sync.WaitGroup
+		)
+
+		aliases, err := s.ts.FindAllTabletAliasesInShard(ctx, keyspace, shard)
+		if err != nil {
+			errMessage := fmt.Sprintf("unable to find tablet aliases in shard %v: %v", shard, err)
+			shardResp.Results = append(shardResp.Results, errMessage)
+			resp.Results = append(resp.Results, errMessage)
+			resp.ResultsByShard[shard] = &shardResp
+			validateVersionKeyspaceResponseMutex.Unlock()
+			continue
+		}
+
+		for _, alias := range aliases {
+			if topoproto.TabletAliasEqual(alias, si.PrimaryAlias) {
+				continue
+			}
+
+			tabletWaitGroup.Add(1)
+			go func(alias *topodatapb.TabletAlias, m *sync.Mutex) {
+				validateShardResponseMutex.Lock()
+				defer validateShardResponseMutex.Unlock()
+				defer tabletWaitGroup.Done()
+				replicaVersion, err := s.GetVersion(ctx, &vtctldatapb.GetVersionRequest{TabletAlias: alias})
+				if err != nil {
+					shardResp.Results = append(shardResp.Results, fmt.Sprintf("unable to get version for tablet %v: %v", alias, err))
+					return
+				}
+
+				if referenceVersion.Version != replicaVersion.Version {
+					shardResp.Results = append(shardResp.Results, fmt.Sprintf("primary %v version %v is different than replica %v version %v", topoproto.TabletAliasString(referenceAlias), referenceVersion, topoproto.TabletAliasString(alias), replicaVersion))
+				}
+			}(alias, &validateShardResponseMutex)
+		}
+
+		tabletWaitGroup.Wait()
+		resp.Results = append(resp.Results, shardResp.Results...)
+		resp.ResultsByShard[shard] = &shardResp
+		validateVersionKeyspaceResponseMutex.Unlock()
+	}
+
+	return &resp, nil
+}
+
+// ValidateVSchema compares the schema of each primary tablet in "keyspace/shards..." to the vschema and errs if there are differences
+func (s *VtctldServer) ValidateVSchema(ctx context.Context, req *vtctldatapb.ValidateVSchemaRequest) (*vtctldatapb.ValidateVSchemaResponse, error) {
+	keyspace := req.Keyspace
+	shards := req.Shards
+	excludeTables := req.ExcludeTables
+	includeViews := req.IncludeViews
+
+	vschm, err := s.ts.GetVSchema(ctx, keyspace)
+	if err != nil {
+		return nil, fmt.Errorf("GetVSchema(%s) failed: %v", keyspace, err)
+	}
+
+	resp := vtctldatapb.ValidateVSchemaResponse{
+		Results:        []string{},
+		ResultsByShard: make(map[string]*vtctldatapb.ValidateShardResponse, len(shards)),
+	}
+
+	var (
+		wg    sync.WaitGroup
+		mutex sync.Mutex
+	)
+
+	wg.Add(len(shards))
+
+	for _, shard := range shards {
+		go func(shard string) {
+			mutex.Lock()
```

Comment:
Locking here effectively removes any benefits we are getting from concurrency, since only one goroutine can execute at a time; this means we pay the scheduler overhead without any upside. I'd suggest removing this, and lock/unlocking in each section where you need to mutate the overall `resp` data structure.